File: c/primes.c
/*
 * @Author:	Jeff Berube
 * @Title:	Primes
 *
 * @Usage	./primes <MAX> <NUMTHREADS>
 *
 * 		<MAX>		Maximum number to calculate primes to
 * 		<NUMTHREADS>	Number of threads to split the workload amongst
 *
 * @Description:	Calculates primes from 1 to MAX using NUMTHREADS threads.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

int max, numthreads;

pthread_mutex_t lock;

/*
 * sieve
 * 
 * Thread function. Parameter is an int between 0 and NUMTHREADS that refers to the interval the
 * thread will operate on.
 *
 */

void *sieve(void *param) {

	/* Variables */
	int flag = 0;
	int threadnum = (int)param ;
	int interval_size = max / numthreads;
	int start = (threadnum * interval_size) + 1;
	int end = (threadnum + 1) * interval_size;
	struct timeval t_start, t_end;

	/* Allocate space to store all primes */
	int *primes = malloc((interval_size / 2) * sizeof(int));
	int index = 0;

	/* If last thread and range doesn't divide nicely by NUMTHREADS */
	if (threadnum == numthreads - 1 && max % interval_size) end += max % numthreads;
	
	/* This is where the magic happens */
	int i, j;	

	/* Start timer for thread runtime */
	gettimeofday(&t_start);

	for (i = start; i < end; ++i) {

		flag = 0;

		for (j = 2; j <= i/2; ++j) {

			if (i % j == 0) {
				flag = 1;
				break;
			}

		}

		if (!flag) {
			
			*(primes + index) = i;
			index++;

		}

	}

	/* End timer for thread runtime */
	gettimeofday(&t_end);

	/* Enter critical output region */
	pthread_mutex_lock(&lock);

	/* Show thread stats */
	printf("\n\nThread: %d \tStart: %d \tEnd: %d \tRuntime: %lu\xC2\xB5s\n", 
			threadnum + 1, start, end, t_end.tv_usec - t_start.tv_usec);
	
	/* Dump primes */
	printf("Primes: ");

	i = 0;
	
	while (i < index) {
	
		printf("%d ", *(primes + i));
		i++;

	}

	pthread_mutex_unlock(&lock);

	/* Exit critical region */

	free(primes);

}

int main (int argc, char *argv[]) {

	/* Arguments */
	max = atoi(argv[1]);
	numthreads = atoi(argv[2]);
	
	pthread_t *threads = malloc(numthreads * sizeof(pthread_t));

	/* Thread variables */
	pthread_t tid;
	pthread_attr_t attr;

	/* Test error conditions */
	if (argc != 3) {
	
		fprintf(stderr, "Usage: primes <max> <numthreads>\n");
		return -1;

	} else if (max < 10) {

		fprintf(stderr, "You must enter a number greater than 10 as the upper bound.\n");
		return -1;

	} else if (numthreads >= max / 2) {

		fprintf(stderr, "Number of threads must be less than half of MAX.\n");
		return -1;		

	} else if (numthreads > 50) {
	
		fprintf(stderr, "Maximum number of threads is 50.");
		return -1;
	}

	/* Init mutex */
	if (pthread_mutex_init(&lock, NULL) != 0) {
	
		printf("Mutex init failed. Try restarting.");
		return -1;

	}

	/* Initiate and create thread */
	pthread_attr_init(&attr);

	int i = 0;

	while (i < numthreads) {
		
		pthread_create(threads + i, &attr, sieve, (void *)i);
		i++;
	}


	i = 0;

	while (i < numthreads) {

		pthread_join(*(threads + i), NULL);
		i++;

	}
	
	pthread_mutex_destroy(&lock);

	printf("\n\n");

	return  0;

}

File: java/Primes.java
/*
 * @Author:	Jeff Berube
 * @Title:	Primes (Java)
 * @Description:	Finds all the primes from 1 to MAX using NUMTHREADS threads.
 *
 * @Usage:	primes <MAX> <NUMTHREADS>
 *
 *		<MAX>		Integer number between 4 and MAXINT (Usually 2^31).
 *				The program will calculate primes between 1 and MAX.
 *
 *		<NUMTHREADS>	Integer number between 1 and 50.
 *				The program will split the workload amongst NUMTHREADS
 *				threads.
 *
 */

public class Primes {

	/* Define class variables */
	public static int max, numthreads;

	public static void main(String[] args) {

		/* Test to see the minimum of arguments is met */
		if (args.length < 2) {

			System.out.println("Invalid number of arguments.");
			System.out.println("Usage: primes <MAX> <NUMTHREADS>");

			return;			

		}

		/* Parse integers out of command line arguments */
		max = Integer.parseInt(args[0]);
		numthreads = Integer.parseInt(args[1]);

		/* Test error conditions before running program */
		if (max < 4) {
			
			System.out.println("MAX must be greater than 4.");
			return;

		} else if (numthreads < 1 || numthreads > 50) {
		
			System.out.println("NUMTHREADS must be between 1 and 50");
			return;

		}

		/* Create variables */
		int i = 0, interval, start, end;
		Thread[] threads = new Thread[numthreads];

		interval = max / numthreads;

		/* Create threads */
		while (i < numthreads) {

			start = (i * interval) + 1;

			if (i == numthreads - 1 && max % numthreads != 0)
				interval = max / numthreads + (max % numthreads);

			end = start + interval - 1;

			threads[i] = new Thread(new Sieve(start, end, i));
			threads[i].start();
	
			i++;

		}

		/* Wait on threads to finish */
		i = 0;

		while (i < numthreads) {
		
			try {

				threads[i++].join();
			
			} catch (InterruptedException ie) {}

		}

		/* One last little output for prettiness */
		System.out.println("\n\n");

	}	

}

class Sieve implements Runnable {

	private int start, end, threadnum, flag = 0, index = 0;
	private int[] primes;
	private long lStartTime, lEndTime, execTime;

	public Sieve(int start, int end, int threadnum) {

		this.start = start;
		this.end = end;
		this.threadnum = threadnum;

		this.primes = new int[(end - start) / 2];

	}

	public void run() {

		/* Store time at start of thread */
		lStartTime = System.currentTimeMillis();

		/* Super magical sieve algorithm */
		int i, j;

		for (i = start; i < end; ++i) {

			flag = 0;

			for (j = 2; j <= i/2; ++j) {

				if (i % j == 0) {

					flag = 1;
					break;

				}

			}	

			/* If number is prime, store in array */
			if (flag == 0) primes[index++] = i;

		}		

		/* Calculate thread runtime */
		execTime = System.currentTimeMillis() - lStartTime;

		/* Output results */
		dumpPrimes();

	}

	/* Dumps all the primes found in the thread to standard output */
	public synchronized void dumpPrimes() {

		synchronized (System.out) {

			System.out.println("\n\nThread: " + threadnum + "\tStart: " 
				+ start + "\tEnd: " + end 
				+ "\tExecution time: " + execTime + "ms");

			System.out.print("Primes: ");

			int i = 0;

			while (i < index) {

				System.out.print(primes[i] + " ");
				i++;

			}

		}

	}


}
